### ISSUE TITLE
dataclass with methods cannot use init=False fields

### ISSUE THREAD

**Issue by toslunar** [pydantic/pydantic#4585]
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

Since pydantic 1.10.2, it fails to parse a builtin dataclass `dataclass.field(init=False)` if it has at least one method.
Similarly to #4498, the cause seems [this change](https://github.com/pydantic/pydantic/pull/4484/files#diff-8ff2ca642bcd607c31ac848396deaf4353b8b4a835473cbca58a22376fa72efaR200) in #4484.

### Example Code

```Python
import dataclasses
import pydantic

@dataclasses.dataclass
class Ok:
    a: int
    b: str = "b"
    c: float = dataclasses.field(init=False)

    def __post_init__(self):
        self.c = float(self.a)

@dataclasses.dataclass
class NotOkSince1_10_2:
    a: int
    b: str = "b"
    c: float = dataclasses.field(init=False)

    def __post_init__(self):
        self.c = float(self.a)

    @pydantic.validator("b")
    def check_b(cls, v):
        if not v:
            raise ValueError("b should not be empty")
        return v

    # ... or with a non-validator method
    # def foo(self):
    #     print("foo")


print(pydantic.parse_obj_as(Ok, {"a": 3}))  # Ok(a=3, b='b', c=3.0)
print(pydantic.parse_obj_as(NotOkSince1_10_2, {"a": 3}))  # TypeError: non-default argument 'c' follows default argument
```

### Python, Pydantic & OS Version

```Text
pydantic version: 1.10.2
            pydantic compiled: True
                 install path: /Users/tos/.pyenv/versions/3.10.0/envs/p10/lib/python3.10/site-packages/pydantic
               python version: 3.10.0 (default, Oct 26 2021, 11:19:32) [Clang 11.0.0 (clang-1100.0.33.16)]
                     platform: macOS-11.7-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
```

### Affected Components

- [ ] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [ ] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [ ] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [X] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [ ] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by samuelcolvin** [comment]
@PrettyWood, any idea on this?

My instinct is that we have to call this a limitation of pydantic v1.10 and fix in v2.

**Comment by PrettyWood** [comment]
This will be fixed in 1.10.3

**Comment by samuelcolvin** [comment]
Closed via #4878
